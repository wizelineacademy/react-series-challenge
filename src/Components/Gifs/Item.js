import React from "react";
import { connect } from "react-redux";
import { ItemStyled, ItemButtonStyled } from "./Item.styled";

const Item = ({
  item,
  favorites: { items: favoriteItems },
  onToggleFavorite
}) => {
  const {
    title,
    images: {
      fixed_width: { url, height, width }
    }
  } = item;

  return (
    <ItemStyled>
      <img src={url} alt={title} width={width} height={height} />
      <ItemButtonStyled
        onClick={() => {
          onToggleFavorite(item);
        }}
        isFavorite={item.id in favoriteItems}
      >
        <i className="far fa-thumbs-up" />
      </ItemButtonStyled>
    </ItemStyled>
  );
};

const mapStateToProps = state => {
  const { favorites } = state;
  return {
    favorites
  };
};

const mapDispatchToProps = dispatch => {
  return {
    onToggleFavorite: item =>
      dispatch({ type: "FAVORITE_TOGGLE", payload: { item } })
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Item);
