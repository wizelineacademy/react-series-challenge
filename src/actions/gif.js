const GIF_CLICK = 'GIF_CLICK'

const clickGif = payload => ({ type: GIF_CLICK, payload })

export default {
    types: {
        GIF_CLICK
    },
    creators: {
        clickGif
    }
}