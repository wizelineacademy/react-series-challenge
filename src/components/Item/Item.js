import React from 'react'
import favIcon from '../../assets/fav.svg'
import noFavIcon from '../../assets/no-fav.svg'
import {
  Item,
  Favorite,
  Image
} from './Item.styled'

const ItemComp = ({item, onClickFav, isFav, onClick}) => {

  const favImage = isFav ? favIcon : noFavIcon
  return(
    <Item>
      <Favorite src={favImage} id="fav-button" onClick={() => onClickFav(item, isFav)} alt="Favorite"/>
      <Image src={item.images.original.webp} alt={item.title} onClick={() => onClick(item)}/>
    </Item>
  )
}

export default ItemComp
