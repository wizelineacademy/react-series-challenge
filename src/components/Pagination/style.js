import styled from 'styled-components'

export const PaginationButton = styled.span`
  color: black;
  background-color: white;
  padding: 8px 16px;
  text-decoration: none;
  transition: background-color .3s;
  border: 1px solid #ddd;
  cursor: pointer;
  margin-top: 30px;
  ${ props => props.active && 
    `background-color: #4CAF50;
    color: white;
    border: 1px solid #4CAF50;
    cursor: default;
    `
  }
`