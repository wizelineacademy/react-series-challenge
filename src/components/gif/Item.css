.gif {
  width: 25%;
  display: inline-block;
}

figure {
  margin: 0;
  text-align: center;
  padding: 0;
}

figure > img {
  width: 100%;
}
