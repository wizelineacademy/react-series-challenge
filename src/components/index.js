import Card from './Card'
import Detail from './Detail'
import GifItem from './GifItem'
import Grid from './Grid'
import Header from './Header'
import Home from './Home'
import SearchBar from './SearchBar'

export {
  Card,
  Detail,
  GifItem,
  Grid,
  Header,
  Home,
  SearchBar
}
