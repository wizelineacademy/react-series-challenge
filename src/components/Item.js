import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { NavLink } from 'react-router-dom';
import { addFavorite, removeFavorite } from '../actions';
import { FavButton, ItemDiv } from './styled';

class Item extends Component {
  handleClick() {
    const {
      addFavorite,
      removeFavorite,
      isFav,
      item
    } = this.props;

    if (isFav) {
      removeFavorite(item.id);
    } else {
      addFavorite(item);
    }
  }

  render() {
    const { item, isFav, imgUrl } = this.props;
    return (
      <ItemDiv>
        <NavLink to={`/Details/${item.id}`}>
          <img src={imgUrl} alt={item.title}/>
        </NavLink>
        <FavButton
          className={isFav ? 'fav' : ''}
          onClick={this.handleClick.bind(this)}
        />
      </ItemDiv>
    );
  }
}

const mapStateToProps = ({ favorites }, { id }) =>
  ({ isFav: !!favorites.items[id] });

const mapDispatchToProps = dispatch =>
  bindActionCreators({ addFavorite, removeFavorite }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Item);
