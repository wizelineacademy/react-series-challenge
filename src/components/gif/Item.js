import React from 'react';
import { Item, FavoriteButton } from './Item.style';

export default ({ data, toggleFavorite, isFavorite }) => {
  return (
    <Item>
      <figure>
        <img src={data.images.original.webp} alt={data.title} />
        <figcaption>
          <FavoriteButton
            favorite={isFavorite}
            onClick={() => toggleFavorite({ gif: data, isFavorite })}
          />
          <p>{data.title}</p>
        </figcaption>
      </figure>
    </Item>
  );
};
