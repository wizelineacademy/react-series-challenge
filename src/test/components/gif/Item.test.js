import React from 'react';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { Item } from '../../../components/gif';
import { FavoriteButton } from '../../../components/gif/Item.style';
import 'jest-styled-components';

describe('Item component', () => {
  test('should render correctly', () => {
    const fakeData = {
      images: {
        original: { webp: '' }
      }
    };
    const component = shallow(<Item data={fakeData} />);

    expect(shallowToJson(component)).toMatchSnapshot();
  });

  test('should be able to toggle favorite gif when favorite button is clicked', () => {
    const toggleFavorite = jest.fn();
    const fakeData = {
      images: {
        original: { webp: '' }
      }
    };
    const component = mount(
      <Item toggleFavorite={toggleFavorite} data={fakeData} />
    );

    component.find('button').simulate('click');
    expect(toggleFavorite).toBeCalled();
  });
});

describe('Item styled component', () => {
  test('should be render favorite icon if received favorite=true', () => {
    const fakeData = {
      images: {
        original: { webp: '' }
      }
    };
    const component = shallow(<Item favorite={true} data={fakeData} />);

    const favoriteButton = shallowToJson(component.find(FavoriteButton));
    expect(favoriteButton).toMatchSnapshot();
  });
});
