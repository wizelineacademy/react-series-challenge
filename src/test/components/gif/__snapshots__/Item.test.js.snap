// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item component should render correctly 1`] = `
<styled.div>
  <figure>
    <img
      src=""
    />
    <figcaption>
      <styled.button
        onClick={[Function]}
      />
      <p />
    </figcaption>
  </figure>
</styled.div>
`;

exports[`Item styled component should be render favorite icon if received favorite=true 1`] = `
<styled.button
  onClick={[Function]}
/>
`;
